### PR TITLE
More API and auto-Login

### PR DESCRIPTION
**Pull request by tagno25**
Added ability to set and get status of lights, chose preset bed position, set sleep number, set and get favorite sleep number, stop bed motion, and stop the pump.

Moved API request to it's own definition and added auto login.

### CODE REVIEW

**Comment by tagno25** [comment]
I am wondering if this will ever get merged and updated on PyPI.  I need to get it updated to fix a bug that stops sleepyq working with home-assistant if a bed only has one sleeper.  And I am working on adding support for bed positions and lighting to home-assistant.  May look into setting the sleep number from home-assistant.

**Comment by CCOSTAN** [comment]
Setting the sleep number from HA would be amazing.  I'd love to have the bed inflate to 100 firm and then have HA set it to the correct sleep number at dusk.  

THAT WOULD BE AMAZING.

File: sleepyq/__init__.py [hunk 1]
```diff
@@ -36,66 +36,208 @@ def __init__(self, data):
...
+        r=self.__makeRequest(url)
         sleepers = [Sleeper(sleeper) for sleeper in r.json()['sleepers']]
         return sleepers
 
     def beds(self):
-        url = 'https://api.sleepiq.sleepnumber.com/rest/bed'
-        r = self._session.get(url)
-        r.raise_for_status()
+        url = '/bed'
+        r=self.__makeRequest(url)
```
Comment (technicalpickles) [inline sleepyq/__init__.py]: If `url` is only used to make the request, I think we don't need the variable, ie `self.__makeRequest('/bed')`.

File: sleepyq/__init__.py [hunk 2]
```diff
@@ -36,66 +36,208 @@ def __init__(self, data):
...
         sleepers_by_id = {sleeper.sleeper_id: sleeper for sleeper in sleepers}
         bed_family_statuses_by_bed_id = {family_status.bed_id: family_status for family_status in family_statuses}
-
         for bed in beds:
             family_status = bed_family_statuses_by_bed_id[bed.bed_id]
-
             for side in ['left', 'right']:
                 sleeper_key = 'sleeper_' + side + '_id'
                 sleeper_id = getattr(bed, sleeper_key)
+                if sleeper_id == "0":
```
Comment (technicalpickles) [inline sleepyq/__init__.py]: Is "0" what happens when you have one sleeper? If so, would you mind including a comment to that effect?

File: sleepyq/__init__.py [hunk 3]
```diff
@@ -36,66 +36,208 @@ def __init__(self, data):
...
-
+        url = '/bed/familyStatus'
+        r=self.__makeRequest(url)
         statuses = [FamilyStatus(status) for status in r.json()['beds']]
         return statuses
+
+    def set_light(self, bedId, light, setting):
+        #
+        # light 1-4
+        ### 1=Right Night Stand
```
Comment (technicalpickles) [inline sleepyq/__init__.py]: Maybe we could use some constants for this? That'd let you call like:

```python
client.set_light(RIGHT_NIGHT_STAND)
```

This applies to a few other places, like `get_light`, and in different contexts (`preset`)

File: sleepyq/__init__.py [hunk 4]
```diff
@@ -36,66 +36,208 @@ def __init__(self, data):
...
+        ### 2=Left Night Stand
+        ### 3=Right Night Light
+        ### 4=Left Night Light
+        # setting 0=off, 1=on
```
Comment (technicalpickles) [inline sleepyq/__init__.py]: If it's just on and off, would it make sense as a boolean? That might map to homeassistant a little better too.

File: sleepyq/__init__.py [hunk 5]
```diff
@@ -36,66 +36,208 @@ def __init__(self, data):
...
+        #
+        if 1 <= light <= 4 and setting in (0, 1):
+            url = '/bed/'+bedId+'/foundation/outlet'
+            data = {'outletId': light, 'setting': setting}
+            r=self.__makeRequest(url, "put", data)
+            return True
+        else:
+            return False
```
Comment (technicalpickles) [inline sleepyq/__init__.py]: If you try to tell it use something not supported at all, is it better to return false, or throw an error? Either could work as long as it's documented.

File: sleepyq/__init__.py [hunk 6]
```diff
@@ -36,66 +36,208 @@ def __init__(self, data):
...
+        self.right = None
+
 class Sleepyq:
     def __init__(self, login, password):
         self._login = login
         self._password = password
         self._session = requests.Session()
+        self._api = "https://api.sleepiq.sleepnumber.com/rest"
+
+    def __makeRequest(self, url, mode="get", data=""):
```
Comment (technicalpickles) [inline sleepyq/__init__.py]: Doesn't python tend to favor snakecase, ie `__make_request`?

File: sleepyq/__init__.py [hunk 7]
```diff
@@ -36,66 +36,208 @@ def __init__(self, data):
...
+        del self._session.params['side']
+        return True
+
+    def set_favsleepnumber(self, bedId, side, setting):
+        #
+        # side "R" or "L"
+        # setting 0-100 (increments of 5)
+        #
+        if setting%5 != 0:
+            return False
```
Comment (technicalpickles) [inline sleepyq/__init__.py]: An error for this with what failed would help inform the user what went wrong. Another option could be rounding to the nearest 5.

It also isn't catching other values, ie negative or more than 100.

File: sleepyq/__init__.py [hunk 8]
```diff
@@ -36,66 +36,208 @@ def __init__(self, data):
...
+        if side.lower() in ('r', 'right'):
+            side = "R"
+        elif side.lower() in ('l', 'left'):
+            side = "L"
+        else:
+            return False
```
Comment (technicalpickles) [inline sleepyq/__init__.py]: Another place that a more specific error coudl be helpful to users of the library, ie raise that left or right is expected.

File: sleepyq/__init__.py [hunk 9]
```diff
@@ -36,66 +36,208 @@ def __init__(self, data):
...
+        r=self.__makeRequest(url, "put", data)
+        return True
+
+    def get_favsleepnumber(self, bedId):
+        url = '/bed/'+bedId+'/sleepNumberFavorite'
+        r=self.__makeRequest(url)
+        favsleepnumber = FavSleepNumber(r.json())
+        return favsleepnumber
+
+    def stopmotion(self, bedId, side):
```
Comment (technicalpickles) [inline sleepyq/__init__.py]: Maybe this should have a dash, ie `stop_motion`?

File: sleepyq/__init__.py [hunk 10]
```diff
@@ -36,66 +36,208 @@ def __init__(self, data):
...
+        elif side.lower() in ('l', 'left'):
+            side = "L"
+        else:
+            return False
+        url = '/bed/'+bedId+'/foundation/motion'
+        data = {"footMotion":1, "headMotion":1, "massageMotion":1, "side":side}
+        r=self.__makeRequest(url, "put", data)
+        return True
+
+    def stoppump(self, bedId):
```
Comment (technicalpickles) [inline sleepyq/__init__.py]: Maybe this should have a dash like `stop_pump`?

File: sleepyq/__init__.py [hunk 11]
```diff
@@ -36,66 +36,208 @@ def __init__(self, data):
...
+    def __makeRequest(self, url, mode="get", data=""):
+        if mode == 'put':
+            r = self._session.put(self._api+url, json=data)
+        else:
+            r = self._session.get(self._api+url)
+        if r.status_code == 401: # HTTP error 401 Unauthorized
+            # Login
+            self.login()
+            # Retry Request
+            if mode == 'put':
```
Comment (technicalpickles) [inline sleepyq/__init__.py]: I wonder if it'd be possible to re-use the logic for retrying? Maybe call the method again, and track how many times it was retried?

File: sleepyq/__init__.py [hunk 12]
```diff
@@ -36,66 +36,208 @@ def __init__(self, data):
...
+                r = self._session.put(self._api+url, json=data)
+            else:
+                r = self._session.get(self._api+url)
+        if r.status_code != 200: # If status code is not 200 OK
+            # Raise error
```
Comment (technicalpickles) [inline sleepyq/__init__.py]: I think `raise_for_status` is clear enough without a comment

**Review by technicalpickles — CHANGES_REQUESTED** [review]
Thanks for putting this together! I've wanted to be able to adjust the firmness for awhile (see https://github.com/technicalpickles/sleepyq/issues/2).

I like this improvement in general, but had some questions and suggestions.

Aside from what is inline, I'd suggest making your change to home-assistant along with this PR. When I wrote the original support, I started with this API and ended up making a bunch of changes to make it a little easier from the HA side of things, so I suspect you'd find that as well. If you have a PR going, please let me know so I can 👀  it too.

Also worth noting my HA server has been down so I can't verify this directly. My bed also doesn't have the lights so I can't validate that either.

**Comment by CCOSTAN** [comment]
Please tag me in the HA PR as well.  I can help test anything you need as I have the sleepIQ and HA running.  @tagno25 @technicalpickles 

Excited for this change!

**Comment by tagno25** [comment]
I have fixed the noted issues and found one or two that I didn't notice before.

It would be nice if someone without lights could get the JSON that is returned when you call get_light, that way the function can be updated to return None if there is no light. 

Also, if someone has a single chamber bed, it would be nice to see if it is reported as the right side or left side (or something else)

File: sleepyq/__init__.py (hunk 6 above)
Comment (tagno25) [reply sleepyq/__init__.py]: changed to snakecase, should I change set_favsleepnumber and get_favsleepnumber as well?

File: sleepyq/__init__.py (hunk 11 above)
Comment (tagno25) [reply sleepyq/__init__.py]: Added ability to count attempts to run requests and limited to 2 tries.  I could make it try a few more times if necessary.

File: sleepyq/__init__.py (hunk 12 above)
Comment (tagno25) [reply sleepyq/__init__.py]: removed that comment and the comments for what the light numbers and presets represent.

File: sleepyq/__init__.py (hunk 1 above)
Comment (tagno25) [reply sleepyq/__init__.py]: I moved all the "url" variables into the __make_request function call

File: sleepyq/__init__.py (hunk 2 above)
Comment (tagno25) [reply sleepyq/__init__.py]: Yes, a sleeper_id of 0 is what is returned by the API when there is no sleeper.  Added comment on that line, may need to be more descriptive

File: sleepyq/__init__.py (hunk 3 above)
Comment (tagno25) [reply sleepyq/__init__.py]: I created some constants, but not sure if I put them in the correct location.

File: sleepyq/__init__.py (hunk 4 above)
Comment (tagno25) [reply sleepyq/__init__.py]: changed to use True/False here as well as the preset speed value

File: sleepyq/__init__.py (hunk 5 above)
Comment (tagno25) [reply sleepyq/__init__.py]: I changed all of the "return False" to "raise ValueError()" describing the reason for the error

File: sleepyq/__init__.py (hunk 7 above)
Comment (tagno25) [reply sleepyq/__init__.py]: Changed to raise ValueError if too small or too large.  Also set the SleepNumber to round to the nearest multiple of 5.  Should it raise an error if it has to be rounded?

File: sleepyq/__init__.py (hunk 8 above)
Comment (tagno25) [reply sleepyq/__init__.py]: I changed all of the "return False" to "raise ValueError()" describing the reason for the error

File: sleepyq/__init__.py (hunk 9 above)
Comment (tagno25) [reply sleepyq/__init__.py]: Changed to stop_motion

File: sleepyq/__init__.py (hunk 10 above)
Comment (tagno25) [reply sleepyq/__init__.py]: Changed to stop_pump

**Review by technicalpickles — APPROVED** [review]
This is looking much better, thanks for applying my feedback ✨ 

> It would be nice if someone without lights could get the JSON that is returned when you call get_light, that way the function can be updated to return None if there is no light.

I don't have lights, so will try to get this for you sometime this week.

**Comment by technicalpickles** [comment]
Hm, I think I was able to pull down the JSON by logging into sleepiq.sleepnumber.com, and hitting the endpoint (`https://prod-api.sleepiq.sleepnumber.com/rest/bed/<some bed id>/foundation/outlet?_k=< some key> `) . I don't see anything that indicates it doesn't support a light though:

```
{"bedId":"redacted","outlet":1,"setting":0,"timer":null}
```

I wonder if there's some other attribute that indicates light support? It might be okay to punt on that in order to get this initial support though.

File: README.rst [hunk 13]
```diff
@@ -19,7 +19,11 @@ To get started using the library, here's the full usage:
     >>> client.login()
     >>> pprint(client.sleepers())
     >>> pprint(client.beds())
-    >>> pprint(bed_family_status())
+    >>> pprint(client.bed_family_status())
+    >>> client.set_lights(bedNumber, lightNumber, setting)
+    >>> pprint(client.get_lights(bedNumber, lightNumber))
+    >>> client.preset(bedNumber, preset, side, speed)
+    >>> client.set_sleepnumber(bedNumber, side, sleepnumber)
```
Comment (technicalpickles) [inline README.rst]: Can we include the other new methods in this output too? Just to keep folks from having to dig in the source just to know what methods are around.

File: sleepyq/__init__.py [hunk 14]
```diff
@@ -36,66 +49,193 @@ def __init__(self, data):
...
+            del self._session.params['outletId']
+            return Status(r.json())
+        else:
+            raise ValueError("Invalid light")
+
+    def preset(self, bedId, preset, side, speed):
+        #
+        # preset 1-6
+        # side "R" or "L"
+        # Speed False=fast, True=slow
```
Comment (technicalpickles) [inline sleepyq/__init__.py]: What is the default speed? Also, I'm not sure if false for fast and true for slow is intuitive, unless the variable is named more sugesting it's going slow (slowSpeed?).

File: sleepyq/__init__.py [hunk 15]
```diff
@@ -36,66 +49,193 @@ def __init__(self, data):
...
+        #
+        if side.lower() in ('r', 'right'):
+            side = "R"
+        elif side.lower() in ('l', 'left'):
+            side = "L"
+        else:
+            raise ValueError("Side mut be one of the following: left, right, L or R")
+        if 1 <= preset <= 6:
```
Comment (technicalpickles) [inline sleepyq/__init__.py]: Maybe it'd be clearer to check if preset in an array of the preset constants?

File: sleepyq/__init__.py (hunk 14 above)
Comment (tagno25) [reply sleepyq/__init__.py]: renamed to slowSpeed, and set as default false.  How should I mark the value as optional in the README? Should I use square brackets around the variable or just "slowSpeed=False"?

File: sleepyq/__init__.py (hunk 15 above)
Comment (tagno25) [reply sleepyq/__init__.py]: changed for presets as well as for lights

File: sleepyq/__init__.py [hunk 16]
```diff
@@ -80,17 +80,20 @@ def __init__(self, login, password):
         self._login = login
         self._password = password
         self._session = requests.Session()
-        self._api = "https://api.sleepiq.sleepnumber.com/rest"
+        self._api = "https://prod-api.sleepiq.sleepnumber.com/rest"
```
Comment (technicalpickles) [inline sleepyq/__init__.py]: Does this mean the existing library is broken now? 🤔

Comment (tagno25) [reply sleepyq/__init__.py]: The old URL still works, but the web interface has changed it's URL. I updated it as a preventive measure.

**Comment by tagno25** [comment]
I just added two more functions one is to request info about he foundation, the other one uses that info to check the features that the foundation has.  I would like to know if the foundation_features function returns the correct info for your bed(s)

File: README.rst [hunk 17]
```diff
@@ -29,6 +29,8 @@ To get started using the library, here's the full usage:
     >>> client.stop_motion(bedNumber, side)
     >>> client.stop_pump(bedNumber)
     >>> pprint(client.foundation_status(bedNumber))
+    >>> pprint(foundation_system(bedNumber))
```
Comment (technicalpickles) [inline README.rst]: Should these be `client.foundation_system(bedNumber)`?

Comment (tagno25) [reply README.rst]: yes, fixed

**Comment by tagno25** [comment]
I just added the home-assistant changes, that I have been working on, to my fork of home-assistant.  It does not currently support the sleep number, favorite sleep number, or favorite position adjustments [https://github.com/tagno25/home-assistant](https://github.com/tagno25/home-assistant)

**Comment by toscano** [comment]
@technicalpickles Is there a reason this isn't merged ? I could use it.

**Comment by tagno25** [comment]
@toscano I was waiting until I had more of the home-assistant code done, but if there are no objections, I can merge it.

**Comment by toscano** [comment]
@tagno25 - The `home-assistant` code is why I wanted it. No rush.

**Comment by technicalpickles** [comment]
@tagno25 👍 to merging. We'll just need a new release with this in it too.

**Comment by tagno25** [comment]
Merging, will work on getting some of the features submitted to home-assistant this week.

**Pull request merged**